File: leetcode/STR/6.ZigZagConversion/6ZigZagConversion.py
'''
字符填充 PAYPALISHIRING
0   p
1   a       x
2   y       x
|   |       x
n-1 s[n-1]
向上填充时，收尾两个字符串不再接受新字符
'''
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        l = len(s)
        sc = [''] * numRows
        
        i = 0
        while i < l:
            for idx in range(numRows):
                if i < l:
                    sc[idx] += s[i]
                    i += 1
            for idx in range(numRows-2, 0, -1):
                if i < l:
                    sc[idx] += s[i]
                    i += 1
        print(sc)
        s = ''.join(sc)
        return s

File: leetcode/STR/6.ZigZagConversion/6ZigZagConversion.cpp
class Solution {
public:
    string convert(string s, int numRows) {
        string res;
        if (numRows == 1) return s;
        
        for (int j=0; j<numRows; j++)
        {
            if (j==0 || j==numRows-1)
                for (int i = j;i<s.size(); i+=(numRows-1)*2)
                    res += s[i];
            
            else
            {
                for (int k = j, i=(numRows-1)*2-j;
                    i<s.size() || k<s.size();
                    i+=(numRows-1)*2, k+=(numRows-1)*2)
                {
                    if (k<s.size()) res+=s[k];
                    if (i<s.size()) res+=s[i];
                }
            }
        }
        return res;
    }
};